datafusion-federation/src/optimize.rs

```diff
@@ -0,0 +1,109 @@
+use datafusion::{
+    common::tree_node::{Transformed, TreeNode, TreeNodeRewriter},
+    error::Result,
+    logical_expr:: LogicalPlan,
+    optimizer::{
+        optimizer::ApplyOrder, push_down_filter, OptimizerConfig, OptimizerContext, OptimizerRule
+    }
+};
+
+pub(crate) fn optimize_plan(plan: LogicalPlan) -> Result<LogicalPlan> {
+    let push_down_filter_rule = push_down_filter::PushDownFilter::new();
+    // `push_down_filter` does not use config so it can be default
+    let optimizer_config = OptimizerContext::default();
+
+    let res = match push_down_filter_rule.apply_order() {
+        Some(apply_order) => plan.rewrite(&mut Rewriter::new(
+            apply_order,
+            &push_down_filter_rule,
+            &optimizer_config,
+        )),
+        None => optimize_plan_node(plan, &push_down_filter_rule, &optimizer_config),
+    };
+
+    Ok(res?.data)
+}
+
+struct Rewriter<'a> {
+    apply_order: ApplyOrder,
+    rule: &'a dyn OptimizerRule,
+    config: &'a dyn OptimizerConfig,
+}
+
+impl<'a> Rewriter<'a> {
+    fn new(
+        apply_order: ApplyOrder,
+        rule: &'a dyn OptimizerRule,
+        config: &'a dyn OptimizerConfig,
+    ) -> Self {
+        Self {
+            apply_order,
+            rule,
+            config,
+        }
+    }
+}
+
+impl<'a> TreeNodeRewriter for Rewriter<'a> {
+    type Node = LogicalPlan;
+
+    fn f_down(&mut self, node: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
+        if self.apply_order == ApplyOrder::TopDown {
+            optimize_plan_node(node, self.rule, self.config)
+        } else {
+            Ok(Transformed::no(node))
+        }
+    }
+
+    fn f_up(&mut self, node: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
+        if self.apply_order == ApplyOrder::BottomUp {
+            optimize_plan_node(node, self.rule, self.config)
+        } else {
+            Ok(Transformed::no(node))
+        }
+    }
+}
+
+fn should_run_rule_for_node(node: &LogicalPlan, _rule: &dyn OptimizerRule) -> bool {
+    // this logic is applicable only for `push_down_filter_rule`; we don't have any other rules
+    if let LogicalPlan::Filter(x) = node {
```

Comment:
This is how we control whether `push_down_filter` should run for plan node. For example, we know that pushing a filter down to a `SubqueryAlias` or `Aggregate` node can transform the plan in a way that we are currently unable to unparse, so we just skip this node.